src/components/legend/style.js

```diff
@@ -435,13 +436,24 @@ module.exports = function style(s, gd) {
     }
 
     function stylePies(d) {
+        stylePieLike(d, this, 'pie');
+    }
+
+    function styleFunnelareas(d) {
+        stylePieLike(d, this, 'funnelarea');
+    }
+
+    function stylePieLike(d, lThis, desiredType) {
         var d0 = d[0];
         var trace = d0.trace;
 
-        var pts = d3.select(this).select('g.legendpoints')
-            .selectAll('path.legendpie')
-            .data(Registry.traceIs(trace, 'pie') && trace.visible ? [d] : []);
-        pts.enter().append('path').classed('legendpie', true)
+        var isVisible = (!desiredType) ? Registry.traceIs(trace, desiredType) :
+            (trace.type === desiredType && trace.visible);
+
+        var pts = d3.select(lThis).select('g.legendpoints')
+            .selectAll('path.legend' + desiredType)
+            .data(isVisible ? [d] : []);
+        pts.enter().append('path').classed('legend' + desiredType, true)
```

Comment:
Starting a new thread regarding  @archmoj 's:

> In case of having a sorted pie with not sorted funnelarea labels may strangely hide other colors.

from https://github.com/plotly/plotly.js/pull/3876#discussion_r285733483 which doesn't _just_ happen via `sort`.

---------

Consider the `funnelarea_pie_colorways` mock:

![image](https://user-images.githubusercontent.com/6675409/58109849-60d2b300-7bbc-11e9-88a2-6052b83a35ce.png)

where the pie and funnelarea traces are set with the same `labels`/`values`. Here the legend draws one item per shared label using the colors of the pie trace (as that trace comes first in `gd.data`).

This is confusing as clicking on the legend items toggles both the pie and funnelarea items associated with the label. 

Potential solutions:

- drop `funnelareacolorway`, make funnelarea and pie traces always share the same colorway (set under `layout.colorway` or `piecolorway`)
- split the legend item square into two halves (in general into N parts), one half with the pie color, the other with funnelarea color

Right now, I think prefer splitting the legend item square. Perhaps this is sufficiently edgy case that we could leave for later though.

Reply:
why does this happen? I thought I saw a mock where the funnel and pie had the same colors. Is this because `piecolorway != funnelcolorway` ?

Reply:
> Is this because `piecolorway != funnelcolorway`

Yes in the case of `funnelarea_pie_colorways`. The same outcome can happen with via `marker.colors` too.

Reply:
hehe ah ok. it's quite an edge-case indeed. I would be fine with punting to another PR, although the split legend thing is a neat idea. Would it solve similar multi-pie/marker-color issues? i.e. two pies, same label, different marker.color?

Reply:
> Would it solve similar multi-pie/marker-color issues? i.e. two pies, same label, different marker.color?

Yes it could.

> I would be fine with punting to another PR, although the split legend thing is a neat idea

I'll open a new issue about it.

Reply:
Now tracked in https://github.com/plotly/plotly.js/issues/3885

Resolved.